gateway-ha/src/main/java/com/lyft/data/gateway/ha/router/PrestoQueueLengthRoutingTable.java

```diff
@@ -0,0 +1,336 @@
+  private void computeWeightsBasedOnQueueLength(ConcurrentHashMap<String,
+      ConcurrentHashMap<String, Integer>> queueLengthMap) {
+    synchronized (lockObject) {
+      int sum = 0;
+      int queueSum = 0;
+      int weight;
+      int numBuckets = 1;
+      int equalDistribution = 0;
+      int smallestQueueLn = 0;
+      int lastButOneQueueLn = 0;
+      int maxQueueLn = 0;
+      int calculatedWtMaxQueue = 0;
+
+      weightedDistributionRouting.clear();
+      routingGroupWeightSum.clear();
+
+      log.debug("Computing Weights for Queue Map " + queueLengthMap.toString());
+
+      for (String routingGroup : queueLengthMap.keySet()) {
+        sum = 0;
+        TreeMap<Integer, String> weightsMap = new TreeMap<>();
+
+        if (queueLengthMap.get(routingGroup).size() == 0) {
+          log.warn("No active clusters in routingGroup : " + routingGroup + ". Continue to "
+              + "process rest of routing table ");
+          continue;
+        } else if (queueLengthMap.get(routingGroup).size() == 1) {
+          log.debug("Routing Group: " + routingGroup + " has only 1 active backend. ");
+          weightedDistributionRouting.put(routingGroup, new TreeMap<Integer, String>() {
+                {
+                  put(MAX_WT, queueLengthMap.get(routingGroup).keys().nextElement());
+                }
+              }
+          );
+          routingGroupWeightSum.put(routingGroup, MAX_WT);
+          continue;
+        }
+
+        Map<String, Integer> sortedByQueueLength = queueLengthMap.get(routingGroup).entrySet()
+            .stream().sorted(Comparator.comparing(Map.Entry::getValue))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                (e1, e2) -> e1, LinkedHashMap::new));
+
+        numBuckets = sortedByQueueLength.size();
+        queueSum = sortedByQueueLength.values().stream().mapToInt(Integer::intValue).sum();
+        equalDistribution = queueSum / numBuckets;
+
+        Object[] queueLengths = sortedByQueueLength.values().toArray();
+        Object[] clusterNames = sortedByQueueLength.keySet().toArray();
+
+        smallestQueueLn = (Integer) queueLengths[0];
+        maxQueueLn = (Integer) queueLengths[queueLengths.length - 1];
+        lastButOneQueueLn = smallestQueueLn;
+        if (queueLengths.length > 2) {
+          lastButOneQueueLn = (Integer) queueLengths[queueLengths.length - 2];
+        }
+
+        /**
+         *  All wts are assigned as a fraction of maxQueueLn. Cluster with maxQueueLn should be
+         *  given least weightage. What this value should be depends on the queueing on the rest of
+         *  the clusters. since the list is sorted, the smallestQueueLn & lastButOneQueueLn give us
+         *  the range. In an ideal situation all clusters should have equals distribution of
+         *  queries, hence that is used as a threshold to check is a cluster queue is over
+         *  provisioned or not.
+         */
+
+        if (maxQueueLn == 0) {
```

Comment:
I'm not sure I understand what each branch of this conditional is doing — it reads to me like this is trying to set a max weight such that all of the weights will sum to a particular value, and establish a discontinuous jump in the weight distribution at the last-but-one-th element?

I think it might be simpler to reduce the weighting to a linear interpolation — pseudocode, but:
```
given queueSum.
for each cluster
  fractionOfQueueSum = cluster.weight() / queueSum
  alpha = 1 - fractionOfQueueSum
  weight = alpha * MAX_WT
  bucket_count = floor(NUM_BUCKETS * weight)
  create_n_sampling_buckets(cluster.name, bucket_count)
```

If you're concerned about the bucket count not exactly matching up, we can either make the routing function do its modular math in terms of `weightedDistributionRouting.size()`, or pad the map with an extra bucket pointing to the least-queued cluster, slightly skewing the distribution.

I believe this is basically what's already happening down on L156-157, but I think perhaps we don't need to do any of this conditional special-casing for small routing groups. If we assign each cluster a simple numeric weight (inversely proportional to its queue depth as a fraction of the whole) it should accomplish the same thing, right?

Reply:
Alex, My initial approach was as you mentioned. Wt would be Inverse of the queueDepth. As I ran my tests I realized that the distribution of queries ( especially when the query volume is low) didnt really work well by excluding the most queue up cluster even when the difference between the least queued Up & most queued Up was significant. 

Just to clarify the the computed weights potentially change every min ( as often as the Active Cluster Monitor) and since the gateway is HA with a fleet of proxy hosts, each proxy host would get a small number of queries.

The way this is designed is to have each of the proxy host to be stateless and agnostic of how the query distribution is happening at a global level. 

To account for the above  I wanted to tweak the weight given to the most queued up cluster. Idea is to set a very negligible wt to the most queued up cluster when there are other lesser util clusters avail. Almost removing it from the mix but not really.

Reply:
@bearcage Experimented with suggestion to use a sqr of the inverse (computed wt) ,but that does not work well for cases with some what even distribution. Ive tried to simplify the math to a large extent, to make the code more readable. I however still need to specially handle some edge cases.